libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1339,62 +1458,46 @@ internal class MapRouteLine(
-            targetDistance: Float,
-            targetPoint: Point
-        ): Double {
-            val linePoints = getReversedCoordinates(line)
-            var runningDistance = 0.0
-            var lastPointIndex = 0
-
-            for (currentIndex in linePoints.indices) {
-                if (currentIndex + 1 < linePoints.size) {
-                    val nextSectionDistance = getDistanceBetweenPoints(
-                        linePoints[currentIndex],
-                        linePoints[currentIndex + 1]
-                    )
-                    if (runningDistance + nextSectionDistance > targetDistance) {
-                        val currentPointToTargetPointDist =
-                            TurfMeasurement.distance(
-                                linePoints[currentIndex],
-                                targetPoint,
-                                TurfConstants.UNIT_METERS
-                            )
-
-                        if (currentPointToTargetPointDist > nextSectionDistance) {
-                            runningDistance += nextSectionDistance
-                            lastPointIndex = currentIndex + 1
-                        } else {
-                            lastPointIndex = currentIndex
-                        }
-                        break
-                    } else {
-                        runningDistance += nextSectionDistance
-                    }
-                } else {
-                    lastPointIndex = currentIndex
-                }
+        fun calculateGranularDistances(points: List<Point>): RouteLineGranularDistances {
+            var distance = 0.0
+            val indexArray = SparseArray<RouteLineDistancesIndex>(points.size)
+            for (i in (points.size - 1) downTo 1) {
+                val curr = points[i]
+                val prev = points[i - 1]
+                distance += calculateDistance(curr, prev)
+                indexArray.append(i - 1, RouteLineDistancesIndex(prev, distance))
             }
+            indexArray.append(
+                points.size - 1,
+                RouteLineDistancesIndex(points[points.size - 1], 0.0)
+            )
+            return RouteLineGranularDistances(distance, indexArray)
+        }
 
-            val distFromTargetToLastIndexPoint = getDistanceBetweenPoints(
-                linePoints[lastPointIndex],
-                targetPoint
+        /**
+         * Calculates the distance between 2 points using
+         * [EPSG:3857 projection](https://epsg.io/3857).
+         * Info in [mapbox-gl-js/issues/9998](https://github.com/mapbox/mapbox-gl-js/issues/9998).
+         */
+        fun calculateDistance(point1: Point, point2: Point): Double {
+            val d = doubleArrayOf(
+                (projectX(point1.longitude()) - projectX(point2.longitude())),
+                (projectY(point1.latitude()) - projectY(point2.latitude()))
             )
+            return sqrt(d[0] * d[0] + d[1] * d[1])
+        }
```

Comment:
Should we expose this from [`mapbox-java`](https://github.com/mapbox/mapbox-java)?

Reply:
Interesting idea, let's track in https://github.com/mapbox/mapbox-java/issues/1208 :+1: